modules/internal/ChapelBase.chpl

```diff
@@ -1007,6 +1007,8 @@ module ChapelBase {
   pragma "down end count fn"
   proc _downEndCount(e: _EndCount, err: unmanaged Error) {
     chpl_save_task_error(e, err);
+    extern proc chpl_comm_task_end(): void;
+    chpl_comm_task_end();
```

Comment:
I saw @mppf's comment about on-stmt bodies.  It seems to me that if there is dirty content in a thread-specific comm cache entry then we do need to flush it at the end of an on-stmt body.  (This because such contents represent comms that are ordered with respect to the initiating task, so we need to make them visible before returning control to the initiating node.)  But I don't think we need to flush any thread-specific unordered comm buffers at the end of an on-stmt body, because such comms are not ordered even with respect to the initiating task.  I think we need to flush both at a sync point, though, such as at the end of a `coforall-stmt`.  Whether we can get this entirely right if we implement it within module code I'm not sure.

It might be nice to add, as building blocks, module functions to be called at the beginnings and ends of task and on-stmt bodies, have the compiler emit calls to those into those statements' body functions, and then the module code could have compile time and/or execution time control flow to decide what needed doing.  But maybe that's for the future.

Reply:
I have added links to this comment and Michael's previous comments in https://github.com/chapel-lang/chapel/issues/12195 -- I will follow up there or create a separate issue to capture this so we don't lose track.